instrumentation/net/http/otelhttpconv/DESIGN.md

```diff
@@ -0,0 +1,283 @@
+	BytesWritten() int64
+
+	// Error contains the last error that occurred while writing
+	Error() error
+}
+```
+
+Note: to keep this document concise, the full implementation is not provided
+here.
+
+### Interfaces
+
+We will provide two public interface that allow interacting with the
+instrumentation package.
+One for client. and one for servers. Implementations can use one, the other or both.
+
+#### The client
+
+```golang
+// Client provides an interface for HTTP client instrumentations to set the
+// proper semantic convention attributes and metrics into their data.
+type Client interface {
+	// RecordError records an error returned by the HTTP request.
+	RecordError(error)
+
+	// Record records the telemetry (spans and metrics) from the provided HTTP request.
+	//
+	// This method does not create a new span. It retrieves the current one from
+	// the context.
+	// It remains the instrumentation's responsibility to start and end spans.
+	Record(ctx context.Context, req *http.Request, w BodyWrapper, cfg ...ClientRecordOption)
+}
+
+// RecordOption applies options to the Record method
+type ClientRecordOption interface{}
+```
+
+#### The Server
+
+```golang
+// Server provides an interface for HTTP server instrumentations to set the
+// proper semantic convention attributes and metrics into their data.
+type Server interface {
+	// Record records the telemetry (spans and metrics) from the provided HTTP
+	// request and response.
+	//
+	// This method does not create a new span. It retrieves the current one from
+	// the context.
+	// It remains the instrumentation's responsibility to start and end spans.
+	Record(ctx context.Context, req *http.Request, w ResponseWrapper, cfg ...ServerRecordOption)
+}
+
+// ServerRecordSpanOption applies options to the RecordSpan method
+type ServerRecordOption interface{}
+```
+
+The `ClientRecordOption` and `ServerRecordOption` functional options allows
+passing optional parameters to be used within the `Record` method, such as the
+HTTP route.
+
+When the data those options provide is not specified, the related span
+attributes will not be set.
+
+### Compatibility
+
+These interfaces are kept relatively simple on purpose, so they can be used as
+foundations if folks wish to extend them with their own actual implementations.
+
+Once stable, we will not be able to change those interfaces anymore (adding,
+renaming or deleting methods), as that would break backwards compatibility with
+implementations.
+
+However, because they remain lightweight, they allow folks to do custom logic,
+such as:
+
+* A proxy that would allow combining multiple implementations together.
+* An implementation that covers unstable semantic conventions.
+* Implementations for older semantic conventions.
+* Implementations that cover custom implementations if a customer doesn't wish to use semconvs.
+
+### Implementations
+
+We will provide one official implementation of the described interfaces.
```

Comment:
What is the point of exposing an interface if we want to provide only one implementation?

Reply:
So that folks can provide their own implementation.
Also, I do mention that we may provide additional implementations later on, for unstable conventions for example.

Reply:
The places where we want to allow passing custom implementations should define the interface (e.g. `otelhttp`). `otelhttpconv` does not accept any custom implementation. 

Reference: https://go.dev/wiki/CodeReviewComments#interfaces